hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieFileGroupReader.java

```diff
@@ -148,27 +148,28 @@ public HoodieFileGroupReader(HoodieReaderContext<T> readerContext, HoodieStorage
   /**
    * Initialize correct record buffer
    */
-  private static FileGroupRecordBuffer getRecordBuffer(HoodieReaderContext readerContext,
-                                                       HoodieTableMetaClient hoodieTableMetaClient,
-                                                       RecordMergeMode recordMergeMode,
-                                                       TypedProperties props,
-                                                       Option<HoodieBaseFile> baseFileOption,
-                                                       boolean hasNoLogFiles,
-                                                       boolean isSkipMerge,
-                                                       boolean shouldUseRecordPosition,
-                                                       HoodieReadStats readStats) {
+  private static <T> FileGroupRecordBuffer<T> getRecordBuffer(HoodieReaderContext<T> readerContext,
+                                                              HoodieTableMetaClient hoodieTableMetaClient,
+                                                              RecordMergeMode recordMergeMode,
+                                                              TypedProperties props,
+                                                              Option<HoodieBaseFile> baseFileOption,
+                                                              boolean hasNoLogFiles,
+                                                              boolean isSkipMerge,
+                                                              boolean shouldUseRecordPosition,
+                                                              HoodieReadStats readStats) {
+    EngineBasedMerger<T> merger = new EngineBasedMerger<>(readerContext, recordMergeMode, hoodieTableMetaClient.getTableConfig(), props);
     if (hasNoLogFiles) {
       return null;
     } else if (isSkipMerge) {
       return new UnmergedFileGroupRecordBuffer<>(
-          readerContext, hoodieTableMetaClient, recordMergeMode, Option.empty(), Option.empty(), props, readStats);
+          readerContext, hoodieTableMetaClient, recordMergeMode, Option.empty(), Option.empty(), props, readStats, merger);
```

Comment:
do we still need to be passing in the merge mode here?

Reply:
The merge mode is used internally for determining whether the ordering field should be set.

Reply:
+1 to @jonvex , why can't we just initialze the `merger` inside each specific record buffer?